File: macclean/cleaners/python_versions.py
import shutil
from pathlib import Path
import click
from rich.console import Console
from rich.table import Table
from rich.panel import Panel
from macclean.core.utils import CleanItem, AnalysisResult, format_size, confirm, dir_size, run_cmd

console = Console()


def _get_pyenv_root(home: Path) -> Path:
    out, code = run_cmd(["pyenv", "root"])
    if code == 0 and out.strip():
        return Path(out.strip())
    return home / ".pyenv"


def _get_pyenv_versions(pyenv_root: Path) -> list[str]:
    versions_dir = pyenv_root / "versions"
    if not versions_dir.exists():
        return []
    return [d.name for d in versions_dir.iterdir() if d.is_dir()]


def _get_active_version(home: Path) -> str:
    out, code = run_cmd(["pyenv", "version-name"])
    if code == 0:
        return out.strip()
    version_file = home / ".pyenv" / "version"
    if version_file.exists():
        return version_file.read_text().strip()
    return ""


def _has_virtualenvs(version: str, pyenv_root: Path) -> bool:
    envs_dir = pyenv_root / "versions" / version / "envs"
    return envs_dir.exists() and any(envs_dir.iterdir())


def analyze(home: Path | None = None) -> AnalysisResult:
    home = home or Path.home()
    result = AnalysisResult()

    if not shutil.which("pyenv"):
        return result

    pyenv_root = _get_pyenv_root(home)
    active = _get_active_version(home)
    versions = _get_pyenv_versions(pyenv_root)

    for version in versions:
        version_path = pyenv_root / "versions" / version
        size = dir_size(version_path)
        has_envs = _has_virtualenvs(version, pyenv_root)
        is_active = version == active
        removable = not is_active and not has_envs
        label = version
        if is_active:
            label += " [active]"
        if has_envs:
            label += " [has virtualenvs]"
        result.items.append(CleanItem(
            label=label,
            path=version_path,
            size_bytes=size,
            removable=removable,
        ))

    return result


def clean(result: AnalysisResult, dry_run: bool = False, yes: bool = False) -> None:
    if not shutil.which("pyenv"):
        console.print("[yellow]pyenv not found — skipping.[/]")
        return

    removable = [i for i in result.items if i.removable]

    table = Table(show_header=True, show_lines=True)
    table.add_column("Python Version")
    table.add_column("Size", justify="right")
    table.add_column("Action")
    for item in result.items:
        action = "[red]Remove[/]" if item.removable else "[dim]Keep[/]"
        table.add_row(item.label, format_size(item.size_bytes), action)

    console.print(Panel(table, title="[bold cyan]Python Versions (pyenv)[/]"))

    if not removable:
        console.print("[green]No unused Python versions found.[/]")
        return

    console.print(f"  Removable: {len(removable)} version(s), [bold]{format_size(sum(i.size_bytes for i in removable))}[/]")

    if dry_run:
        return
    if not yes and not confirm(f"Uninstall {len(removable)} unused Python version(s)?"):
        return

    for item in removable:
        version = item.path.name
        out, code = run_cmd(["pyenv", "uninstall", "-f", version], timeout=60)
        if code == 0:
            console.print(f"  [green]+[/] Removed Python {version}")
        else:
            console.print(f"  [yellow]![/] {version}: {out}")


@click.command()
@click.pass_context
def cmd(ctx):
    """Remove unused pyenv Python versions."""
    result = analyze()
    clean(result, dry_run=ctx.obj["dry_run"], yes=ctx.obj["yes"])

File: src/cleaners/python_versions.rs
use std::path::PathBuf;
use anyhow::Result;
use crate::core::{AnalysisResult, Cleaner, CleanItem};
use crate::core::fs::dir_size;
use crate::core::cmd::run_cmd;
use crate::ui;

pub struct PythonVersionsCleaner;

impl Cleaner for PythonVersionsCleaner {
    fn name(&self) -> &str { "python" }
    fn display_name(&self) -> &str { "Python Versions (pyenv)" }

    fn analyze(&self) -> Result<AnalysisResult> {
        let mut result = AnalysisResult::default();

        if !run_cmd(&["which", "pyenv"]).success() {
            return Ok(result);
        }

        // Get pyenv root
        let root_r = run_cmd(&["pyenv", "root"]);
        let pyenv_root = if root_r.success() {
            PathBuf::from(root_r.output.trim())
        } else {
            dirs::home_dir()
                .unwrap_or_else(|| PathBuf::from("/tmp"))
                .join(".pyenv")
        };

        let versions_dir = pyenv_root.join("versions");
        if !versions_dir.exists() {
            return Ok(result);
        }

        // Get active version
        let active_r = run_cmd(&["pyenv", "version-name"]);
        let active = active_r.output.trim().to_string();

        let entries = match std::fs::read_dir(&versions_dir) {
            Ok(e) => e,
            Err(_) => return Ok(result),
        };

        for entry in entries.filter_map(|e| e.ok()) {
            let path = entry.path();
            if !path.is_dir() { continue; }

            let version_name = path
                .file_name()
                .map(|n| n.to_string_lossy().to_string())
                .unwrap_or_default();

            let is_active = version_name == active;

            // Check if it has virtualenvs
            let envs_dir = path.join("envs");
            let has_envs = envs_dir.exists()
                && std::fs::read_dir(&envs_dir)
                    .map(|mut d| d.next().is_some())
                    .unwrap_or(false);

            let mut label = version_name.clone();
            if is_active { label.push_str(" [active]"); }
            if has_envs { label.push_str(" [has virtualenvs]"); }

            let removable = !is_active && !has_envs;
            let size = dir_size(&path);

            result.items.push(CleanItem {
                label,
                path,
                size_bytes: size,
                removable,
            });
        }

        Ok(result)
    }

    fn clean(&self, result: &AnalysisResult, dry_run: bool, yes: bool) -> Result<()> {
        if result.items.is_empty() {
            println!("No pyenv Python versions found.");
            return Ok(());
        }

        ui::print_analysis("Python Versions (pyenv)", &result.items);

        let removable: Vec<_> = result.items.iter().filter(|i| i.removable).collect();
        if removable.is_empty() {
            println!("No unused Python versions found.");
            return Ok(());
        }

        if dry_run { return Ok(()); }
        if !yes && !ui::confirm(
            &format!("Uninstall {} unused Python version(s)?", removable.len()),
            false,
        )? {
            return Ok(());
        }

        for item in &removable {
            // Extract the plain version name (strip any suffix tags we added)
            let version_name = item.label
                .split_whitespace()
                .next()
                .unwrap_or(&item.label);

            let r = run_cmd(&["pyenv", "uninstall", "-f", version_name]);
            if r.success() {
                ui::print_ok(&format!("Uninstalled Python {}", version_name));
            } else {
                ui::print_warn(&format!("{}: {}", version_name, &r.output[..r.output.len().min(200)]));
            }
        }
        Ok(())
    }
}
